File: project6_lcm_mes/src/api/routes/ai.py
from fastapi import APIRouter, Depends
from src.ai.models import (
    DefectPredictionRequest,
    DefectPredictionResponse,
    EquipmentHealthRequest,
    EquipmentHealthResponse,
    RepairDecisionRequest,
    RepairDecisionResponse,
    YieldPredictionRequest,
    YieldPredictionResponse,
    SchedulingOptimizationRequest,
    SchedulingOptimizationResponse
)
from src.ai.service import AIService
from src.api.dependencies import get_current_user

router = APIRouter(prefix="/api/v1/ai", tags=["AI"])


@router.post("/defect-prediction", response_model=DefectPredictionResponse)
def predict_defect(
    request: DefectPredictionRequest,
    current_user: dict = Depends(get_current_user)
):
    return AIService.predict_defect(
        product_code=request.product_code,
        process_step=request.process_step,
        equipment_id=request.equipment_id,
        operator_id=request.operator_id,
        environmental_data=request.environmental_data
    )


@router.post("/equipment-health", response_model=EquipmentHealthResponse)
def get_equipment_health(
    request: EquipmentHealthRequest,
    current_user: dict = Depends(get_current_user)
):
    return AIService.predict_equipment_health(
        equipment_id=request.equipment_id,
        look_ahead_days=request.look_ahead_days
    )


@router.post("/repair-decision", response_model=RepairDecisionResponse)
def make_repair_decision(
    request: RepairDecisionRequest,
    current_user: dict = Depends(get_current_user)
):
    return AIService.make_repair_decision(
        product_code=request.product_code,
        defect_type=request.defect_type,
        defect_severity=request.defect_severity,
        repair_history=request.repair_history
    )


@router.post("/yield-prediction", response_model=YieldPredictionResponse)
def predict_yield(
    request: YieldPredictionRequest,
    current_user: dict = Depends(get_current_user)
):
    return AIService.predict_yield(
        product_type=request.product_type,
        batch_size=request.batch_size,
        process_parameters=request.process_parameters
    )


@router.post("/scheduling-optimization", response_model=SchedulingOptimizationResponse)
def optimize_scheduling(
    request: SchedulingOptimizationRequest,
    current_user: dict = Depends(get_current_user)
):
    return AIService.optimize_scheduling(
        order_priority=request.order_priority,
        resource_constraints=request.resource_constraints
    )


@router.get("/defect-trends")
def get_defect_trends(
    hours: int = 24,
    current_user: dict = Depends(get_current_user)
):
    return AIService.analyze_defect_trends(hours=hours)


@router.get("/quality-report")
def get_quality_report(
    product_type: str = "all",
    current_user: dict = Depends(get_current_user)
):
    return AIService.generate_quality_report(product_type=product_type)


@router.get("/dashboard")
def get_ai_dashboard(
    current_user: dict = Depends(get_current_user)
):
    return {
        "defect_trends": AIService.analyze_defect_trends(24),
        "quality_report": AIService.generate_quality_report(),
        "yield_forecast": {
            "today": AIService.predict_yield("LCM", 1000),
            "week": {"predicted_yield": round(0.945, 3), "trend": "stable"}
        },
        "equipment_alerts": [
            AIService.predict_equipment_health(1, 7),
            AIService.predict_equipment_health(2, 7),
            AIService.predict_equipment_health(3, 7)
        ]
    }

File: project6_lcm_mes/frontend/src/api/ai.ts
import api from './client'

export interface DefectPredictionRequest {
  product_code: string
  process_step: string
  equipment_id?: number
  operator_id?: number
  environmental_data?: Record<string, number>
}

export interface DefectPredictionResponse {
  product_code: string
  defect_probability: number
  defect_type?: string
  confidence: number
  recommendations: string[]
}

export interface EquipmentHealthRequest {
  equipment_id: number
  look_ahead_days?: number
}

export interface EquipmentHealthResponse {
  equipment_id: number
  equipment_name: string
  health_score: number
  risk_level: string
  predicted_failure_date?: string
  recommended_actions: string[]
}

export interface RepairDecisionRequest {
  product_code: string
  defect_type: string
  defect_severity: string
  repair_history?: Record<string, unknown>[]
}

export interface RepairDecisionResponse {
  product_code: string
  should_repair: boolean
  repair_cost_estimate: number
  repair_time_estimate: number
  alternative_action?: string
  confidence: number
}

export interface YieldPredictionRequest {
  product_type: string
  batch_size: number
  process_parameters?: Record<string, number>
}

export interface YieldPredictionResponse {
  product_type: string
  predicted_yield: number
  confidence: number
  key_factors: string[]
  optimization_suggestions: string[]
}

export interface SchedulingOptimizationRequest {
  order_priority?: string
  resource_constraints?: Record<string, unknown>
}

export interface SchedulingOptimizationResponse {
  optimized_schedule: Record<string, unknown>[]
  expected_throughput: number
  bottleneck_analysis: string[]
}

export interface DefectTrendData {
  hour: string
  defect_count: number
  defect_rate: number
}

export interface DefectTrendsResponse {
  trend_data: DefectTrendData[]
  top_defects: { type: string; count: number; rate: number }[]
  overall_defect_rate: number
  anomaly_detection: { time: string; type: string; value: string }[]
}

export interface QualityReportResponse {
  report_period: string
  total_units: number
  good_units: number
  defect_units: number
  yield_rate: number
  defect_distribution: { type: string; count: number; percentage: number }[]
  recommendations: string[]
  ai_summary: string
}

export interface AIDashboardResponse {
  defect_trends: DefectTrendsResponse
  quality_report: QualityReportResponse
  yield_forecast: {
    today: YieldPredictionResponse
    week: { predicted_yield: number; trend: string }
  }
  equipment_alerts: EquipmentHealthResponse[]
}

export const aiApi = {
  predictDefect: (data: DefectPredictionRequest) =>
    api.post<DefectPredictionResponse>('/api/v1/ai/defect-prediction', data),

  getEquipmentHealth: (data: EquipmentHealthRequest) =>
    api.post<EquipmentHealthResponse>('/api/v1/ai/equipment-health', data),

  makeRepairDecision: (data: RepairDecisionRequest) =>
    api.post<RepairDecisionResponse>('/api/v1/ai/repair-decision', data),

  predictYield: (data: YieldPredictionRequest) =>
    api.post<YieldPredictionResponse>('/api/v1/ai/yield-prediction', data),

  optimizeScheduling: (data?: SchedulingOptimizationRequest) =>
    api.post<SchedulingOptimizationResponse>('/api/v1/ai/scheduling-optimization', data || {}),

  getDefectTrends: (hours?: number) =>
    api.get<DefectTrendsResponse>('/api/v1/ai/defect-trends', { params: { hours } }),

  getQualityReport: (productType?: string) =>
    api.get<QualityReportResponse>('/api/v1/ai/quality-report', { params: { product_type: productType } }),

  getAIDashboard: () => api.get<AIDashboardResponse>('/api/v1/ai/dashboard')
}